### PR TITLE
801 reset password

### LINKED ISSUE #801: 🚀 [Feature] Reset password
## :rocket: Feature

Reset your password and get a mail to login again.

#### User story
```gherkin
As a user
I want to have a button to send me a reset password mail
To be able to login again in case I forgot my password
```

### Design & Layout

Screenshot - 2019-06-12T120331 736

### PR DESCRIPTION
**Pull request by roschaefer**
## 🍰 Pullrequest
Implement a basic password reset flow. I did not follow a security best practice here: You should ask the user for her username. That way, the account is safe even if an attacker has access to the email inbox. Reason for not implementing was that our users cannot choose their own slug at the moment. Therefore I believe we would make the whole flow pretty unusable if we ask for sth. that the user never changed and probably does not remember.

### Issues

- Close #801

### Todo

- [X] None

### CODE REVIEW

**Comment by roschaefer** [comment]
This is still a little bit work-in-progress but I think this is the moment when I can ask you for a first review.

File: webapp/locales/de.json [hunk 1]
```diff
@@ -8,11 +8,32 @@
...
+      "submitted": "Eine E-Mail mit weiteren Instruktionen wurde verschickt an <b>{email}</b>"
+    }
+  },
+  "verify-code": {
+    "form": {
+      "code": "Code eingeben",
+      "description": "Öffne Deine E-Mail Postfach und gib den Code ein, den wir geschickt haben.",
+      "next": "Weiter",
+      "change-password":{
+        "success": "Änderung des Passworts war erfolgreich",
```
Comment (Tirokk) [inline webapp/locales/de.json]: `… war erfolgreich!` we usually put a `!` behind success texts I have the impression.

File: webapp/locales/en.json [hunk 2]
```diff
@@ -8,11 +8,32 @@
...
+      "submitted": "A mail with further instruction has been sent to <b>{email}</b>"
+    }
+  },
+  "verify-code": {
+    "form": {
+      "code": "Enter your code",
+      "description": "Open your inbox and enter the code that we've sent to you.",
+      "next": "Continue",
+      "change-password": {
+        "success": "Changing your password was successful",
```
Comment (Tirokk) [inline webapp/locales/en.json]: `… successful!` we usually put a ! behind success texts I have the impression.

File: backend/src/schema/resolvers/passwordReset.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      RETURN r
+      `
+      let transactionRes = await session.run(cypher, { stillValid, email, code, newHashedPassword })
+      const [reset] = transactionRes.records.map(record => record.get('r'))
+      const result = !!(reset && reset.properties.usedAt)
+      session.close()
+      return result
+    },
+  },
+}
```
[the remark this answers is not in the document]
Comment (roschaefer) [reply backend/src/schema/resolvers/passwordReset.js]: yes

File: backend/src/schema/resolvers/passwordReset.spec.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      password: '1234',
+    })
+  })
+
+  afterEach(async () => {
+    await factory.cleanDatabase()
+  })
+
+  describe('requestPasswordReset', () => {
+    const mutation = `mutation($email: String!) { requestPasswordReset(email: $email) }`
```
[the remark this answers is not in the document]
Comment (roschaefer) [reply backend/src/schema/resolvers/passwordReset.spec.js]: In tests? I usually don't do that - by looking nicer you mean the editor is doing syntax highlighting?

File: backend/src/schema/resolvers/passwordReset/emailTemplates.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  actionUrl.searchParams.set('email', email)
+
+  return {
+    to: email,
+    subject,
+    text: `
+Hi ${name}!
+
+You recently requested to reset your password for your Human Connection account.
+Use the button below to reset it. This password reset is only valid for the next
```
[the remark this answers is not in the document]
Comment (roschaefer) [reply backend/src/schema/resolvers/passwordReset/emailTemplates.js]: Yes, you're right

File: webapp/locales/de.json (hunk 1 above)
Comment (roschaefer) [reply webapp/locales/de.json]: :+1:

**Comment by roschaefer** [comment]
@Tirokk does `yarn install --ignore-engines` resolve the problem? I have never seen that output on my machine and believe it's unrelated.

File: backend/src/schema/resolvers/passwordReset.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    },
+    resetPassword: async (_, { email, code, newPassword }, { driver }) => {
+      const session = driver.session()
+      const stillValid = new Date()
+      stillValid.setDate(stillValid.getDate() - 1)
+      const newHashedPassword = await bcrypt.hashSync(newPassword, 10)
+      const cypher = `
+      MATCH (pr:PasswordReset {code: $code})
+      MATCH (u:User {email: $email})-[:REQUESTED]->(pr)
+      WHERE duration.between(pr.issuedAt, datetime()).days <= 0 AND pr.usedAt IS NULL
```
Comment (mattwr18) [inline backend/src/schema/resolvers/passwordReset.js]: Maybe I'm missing something... the duration between when the password reset was issued and now in days is `less` than or equal to 0? shouldn't it be 1??

Comment (roschaefer) [reply backend/src/schema/resolvers/passwordReset.js]: The duration will be a full day. So `<=1` would mean 1 day + 23 hours + 59 minutes + 59 seconds.

File: backend/src/schema/resolvers/passwordReset.spec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Tirokk) [reply backend/src/schema/resolvers/passwordReset.spec.js]: In tests! Yes!
Test should be good readable as well, or?

[the remark this answers is not in the document]
Comment (mattwr18) [reply backend/src/schema/resolvers/passwordReset.spec.js]: I think the point is that @roschaefer doesn't use VS code

File: backend/src/schema/resolvers/passwordReset.js (hunk 6 above)
Comment (mattwr18) [reply backend/src/schema/resolvers/passwordReset.js]: Sorry, maybe I am being dense, but how is less than or equal to 1 mean 1+?

**Pull request merged**